osquery/tables/events/windows/powershell_events.cpp

```diff
@@ -0,0 +1,198 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <cmath>
+
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+
+#include <osquery/config.h>
+#include <osquery/core.h>
+#include <osquery/database.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/config/parsers/feature_vectors.h"
+#include "osquery/core/conversions.h"
+#include "osquery/core/json.h"
+#include "osquery/core/windows/wmi.h"
+#include "osquery/events/windows/windows_event_log.h"
+#include "osquery/filesystem/fileops.h"
+
+namespace pt = boost::property_tree;
+
+namespace osquery {
+
+const std::string kPowershellDomain{"powershell_script_block_logs"};
+const std::string kScriptBlockPrefix{"script_block."};
+const std::string kScriptBlocksReceivedSuffix{".blocks_received"};
+const std::wstring kPowershellEventsChannel{
+    L"microsoft-windows-powershell/operational"};
+
+const int kScriptBlockLoggingEid{4104};
+const int kCharFreqVectorLen{255};
+
+void parseTree(const pt::ptree& tree, std::map<std::string, std::string>& res);
+
+class PowershellEventSubscriber
+    : public EventSubscriber<WindowsEventLogEventPublisher> {
+ public:
+  Status init() override {
+    auto wc = createSubscriptionContext();
+    wc->sources.insert(kPowershellEventsChannel);
+
+    subscribe(&PowershellEventSubscriber::Callback, wc);
+    return Status();
+  }
+
+  Status Callback(const ECRef& ec, const SCRef& sc);
+
+  void addScriptResult(Row& results);
+};
+
+REGISTER(PowershellEventSubscriber, "event_subscriber", "powershell_events");
+
+inline double cosineSimilarity(std::vector<double>& scriptFreqs,
+                               std::vector<double>& globalFreqs) {
+  auto dot = 0.0;
+  auto mag1 = 0.0;
+  auto mag2 = 0.0;
+
+  for (size_t i = 0; i < globalFreqs.size(); i++) {
+    dot += scriptFreqs[i] * globalFreqs[i];
+    mag1 += scriptFreqs[i] * scriptFreqs[i];
+    mag2 += globalFreqs[i] * globalFreqs[i];
+  }
+
+  mag1 = sqrt(mag1);
+  mag2 = sqrt(mag2);
+
+  return dot / (mag1 * mag2);
+}
+
+void PowershellEventSubscriber::addScriptResult(Row& results) {
+  Row r;
+  r["time"] = results["time"];
+  r["datetime"] = results["datetime"];
+  r["script_block_id"] = results["ScriptBlockId"];
+  r["script_block_count"] = INTEGER(results["MessageTotal"]);
+  r["script_name"] = results["Name"];
+  r["script_path"] = results["Path"];
+  std::string scriptText = results["ScriptBlockText"];
```

Comment:
Let's be the change we wish to see and use `script_text`. But also so script_text and ScriptBlockText contain the same data in the added row?